Single Product Template: Add compatibility layer

gigitux opened the pull request:
<!-- Start by describing the changes made in this Pull Request, and the reason for such changes. -->
This PR adds a compatibility layer for the Single Product Template.

First of all, I created an abstract class called `AbstractTemplateCompatibility`. This way, we can easily reuse the same interface for all the templates that require a compatibility layer. After I created the `SingleProductTemplateClass`, I moved all the logic about the div wrapper (you can see the dedicated PR here #8364) inside the new file. You can check this refactor work in commit 5a8c13f.

After that, I worked on adding the support for the hooks.

Since the classic Single Product template is opinionated, we found a "tricky" solution to support some hooks ([source code](https://github.com/woocommerce/woocommerce-blocks/blob/d715fa3ed60c7d3ea7c0c85924fcac4e113a502d/src/Templates/SingleProductTemplateCompatibility.php/#L53-L129) - [source code](https://github.com/woocommerce/woocommerce-blocks/blob/d715fa3ed60c7d3ea7c0c85924fcac4e113a502d/src/Templates/SingleProductTemplateCompatibility.php/#L275-L328)).

The first block inside the Single Product template has the support for those hooks:
- woocommerce_before_single_product
- woocommerce_before_single_product_summary
- woocommerce_single_product_summary

Furthermore the `Product Details` block will have support for the hooks ([source code](https://github.com/woocommerce/woocommerce-blocks/blob/d715fa3ed60c7d3ea7c0c85924fcac4e113a502d/src/Templates/SingleProductTemplateCompatibility.php/#L184-L196):

- woocommerce_share
- woocommerce_after_single_product_summary

The last block has the support for this hook:
- woocommerce_after_single_product


Before releasing the blockified Single Product Template, we have to write a dev note for this compatibility layer to highlight the caveats of the layers.

⚠️ Also, in the PR missing the support for the hooks:
- woocommerce_share
- woocommerce_product_meta_end

I will create a follow-up PR for them 👍 

<!-- Reference any related issues or PRs here -->

Fixes #8054

### Testing

#### Automated Tests
* [ ] Changes in this PR are covered by Automated Tests.
  * [ ] Unit tests
  * [ ] E2E tests

#### User Facing Testing

<!-- Write these steps from the perspective of a "user" (merchant) familiar with WooCommerce. No need to spell out the steps for common setup scenarios (eg. "Create a product"), but do be specific about the thing being tested. Include screenshots demonstrating expectations where that will be helpful. -->

1. Edit the Single Product Template.
2. Remove the Classic Product Block.
3. Add Product Image Gallery, Add to Cart Form, and Product Details block.
5. Install the Code Snippets plugin.
6. Create a new snippet and copy-paste [this gist](https://gist.github.com/gigitux/b99b823b69dba7a615b82b4200ac5fbb).
7. Be sure that all the hooks work correctly.

(you can check the list of hook on [this page](https://www.businessbloomer.com/woocommerce-visual-hook-guide-single-product-page/#tab-additional_information))

* [x] Do not include in the Testing Notes <!-- Check this box if this PR can't be tested by users (ie: it doesn't include user-facing changes or it can't be tested without manually modifying the code). -->

### WooCommerce Visibility

<!-- Check this [this doc](../docs/blocks/feature-flags-and-experimental-interfaces.md) to see if the change is visible in WC core or not (part of the feature plugin or experimental)-->

* [x] WooCommerce Core
* [ ] Feature plugin
* [ ] Experimental

thealexandrelara committed:
Add minimum structure for Single Product Details block

gigitux committed:
Add Product Image Gallery #8233

Add Product Image Gallery

gigitux committed:
Merge branch 'trunk' of https://github.com/woocommerce/woocommerce-blocks into add/8233-product-image-gallery-block

thealexandrelara committed:
Add tests for Single Product Details block

nefeline committed:
Add the initial basis for the Add to Cart button

nefeline committed:
Trigger the single product add to cart action for each product type.

thealexandrelara committed:
wip: create block structure and add initial styles

thealexandrelara committed:
Add block details to the SingleProductDetails.php file

nefeline committed:
Merge branch 'trunk' into add/add-to-cart-block

nefeline committed:
Rename the block from add-to-cart-button to add-to-cart-form

nefeline committed:
Update to use the cart icon.

nefeline committed:
Implement the skeleton for the editor preview.

thealexandrelara committed:
Render tabs title with empty content

thealexandrelara committed:
Use woocommerce_output_product_data_tabs function to retrieve tabs data

nefeline committed:
Merge branch 'trunk' into add/add-to-cart-block

nefeline committed:
Update styles and add Notice for the display in the Editor.

nefeline committed:
Update CSS.

gigitux committed:
Merge branch 'trunk' of https://github.com/woocommerce/woocommerce-blocks into add/8233-product-image-gallery-block

nefeline committed:
Add base tests for the new Add to Cart Form component.

gigitux committed:
Add Product Image Gallery block

gigitux committed:
Merge branch 'trunk' of https://github.com/woocommerce/woocommerce-blocks into add/8233-product-image-gallery-block

gigitux committed:
remove support global styles

gigitux committed:
remove support global styles

gigitux committed:
Merge branch 'add/8233-product-image-gallery-block' of https://github.com/woocommerce/woocommerce-blocks into add/8233-product-image-gallery-block

gigitux committed:
Merge branch 'trunk' into add/8233-product-image-gallery-block

nefeline committed:
Update the button CSS.

thealexandrelara committed:
Remove customizations for the Single Product Details block

nefeline committed:
Update styles for the cart form.

nefeline committed:
update td style.

thealexandrelara committed:
Merge branch 'trunk' into feat/add-single-product-details-block

nefeline committed:
Update divs and CSS.

nefeline committed:
Merge branch 'trunk' into add/add-to-cart-block

nefeline committed:
Use conventional input instead of the experimental InputControl

gigitux committed:
address CSS feedback

gigitux committed:
Merge branch 'add/8233-product-image-gallery-block' of https://github.com/woocommerce/woocommerce-blocks into add/8233-product-image-gallery-block

gigitux committed:
add support for the custom classname

gigitux committed:
remove save function

thealexandrelara committed:
Remove unnecessary console.log from the Edit.tsx file

thealexandrelara committed:
Remove block classname from block wrapper

thealexandrelara committed:
Remove unnecessary WooCommerce tabs filter from the BlockTemplatesController

thealexandrelara committed:
Merge branch 'trunk' into feat/add-single-product-details-block

thealexandrelara committed:
Remove attributes property from the block registration

thealexandrelara committed:
Remove isExperimental flag for the Single Product Details block

thealexandrelara committed:
Remove get_classes_and_styles_by_attributes method from SingleProductDetails block

thealexandrelara committed:
Prevent Single Product Details block from apppearing in Pages or Posts

gigitux committed:
add second parameter to the subscribe function

nefeline committed:
Implement the new design and copy provided for the editor.

nefeline committed:
Make the notice compatible with dark themes.

nefeline committed:
Some additional CSS tweaks

nefeline committed:
adjust the padding for the input

gigitux committed:
wrap the Single Product Template in a div with the product class

thealexandrelara committed:
Fix PHP Coding Standards warnings

gigitux committed:
improve logic and increase coverage of unit test

nefeline committed:
Merge branch 'trunk' into add/add-to-cart-block

gigitux committed:
improve logic and increase coverage of unit test

gigitux committed:
Merge branch 'trunk' into fix/8314-blockifying-single-product-template-investigate-how-to-apply-wc-core-styles-to-product-template

gigitux committed:
fix test

gigitux committed:
format HTML

gigitux committed:
fix edge case

gigitux committed:
Merge branch 'trunk' of https://github.com/woocommerce/woocommerce-blocks into add/8233-product-image-gallery-block

gigitux committed:
Merge branch 'trunk' of https://github.com/woocommerce/woocommerce-blocks into add/8233-product-image-gallery-block

gigitux committed:
update @types/wordpress__data package

gigitux committed:
update placeholder, icon and description

gigitux committed:
update tsconfig

gigitux committed:
Merge branch 'trunk' into feat/add-single-product-details-block

gigitux committed:
update block name

gigitux committed:
Merge branch 'feat/add-single-product-details-block' of https://github.com/woocommerce/woocommerce-blocks into feat/add-single-product-details-block

gigitux committed:
fix SCSS linter error

gigitux committed:
Merge branch 'add/add-to-cart-block' of https://github.com/woocommerce/woocommerce-blocks into add/compatibility-single-product-block

gigitux committed:
Merge branch 'add/8233-product-image-gallery-block' into add/compatibility-single-product-block

gigitux committed:
Merge branch 'feat/add-single-product-details-block' into add/compatibility-single-product-block

gigitux committed:
Merge branch 'trunk' of https://github.com/woocommerce/woocommerce-blocks into fix/8314-blockifying-single-product-template-investigate-how-to-apply-wc-core-styles-to-product-template

gigitux committed:
address feedback

gigitux committed:
Merge branch 'trunk' of https://github.com/woocommerce/woocommerce-blocks into fix/8314-blockifying-single-product-template-investigate-how-to-apply-wc-core-styles-to-product-template

gigitux committed:
Merge branch 'fix/8314-blockifying-single-product-template-investigate-how-to-apply-wc-core-styles-to-product-template' of https://github.com/woocommerce/woocommerce-blocks into add/compatibility-single-product-block

gigitux committed:
create SingleProductTemplateCompatibility class

gigitux committed:
Merge branch 'trunk' of https://github.com/woocommerce/woocommerce-blocks into add/compatibility-single-product-block

gigitux committed:
Add Hooks compatibility

gigitux committed:
Merge branch 'trunk' of https://github.com/woocommerce/woocommerce-blocks into add/compatibility-single-product-block

gigitux committed:
remove not used file

gigitux committed:
Merge branch 'trunk' of https://github.com/woocommerce/woocommerce-blocks into add/compatibility-single-product-block

gigitux committed:
remove not used files

gigitux committed:
Add compatibility layer for the Single Product template

gigitux committed:
Merge branch 'trunk' of https://github.com/woocommerce/woocommerce-blocks into add/compatibility-single-product-block

gigitux commented:
I added as reviewers multiple folks because a lot of folks are involved/reviewed the code that I updated.

dinhtungdu commented on src/Templates/AbstractTemplateCompatibility.php:
```diff
@@ -0,0 +1,157 @@
+ *
+ * @internal
+ */
+abstract class AbstractTemplateCompatibility {
```
To be honest, I'm not a fan of inheritance, what do you think about using `trait` instead?

gigitux replied:
I didn't know about trait keyword. I will check it

nefeline replied:
+1 on using a trait: they are the perfect use-case for classes designed for reuse.

gigitux replied:
Since this PR is already big, I plan to merge this when you approve and, after that, in another PR, rename `BlockTemplatesCompatibility` to `ArchiveProductTemplateCompatibility` and convert this class to a trait.

dinhtungdu replied:
After consideration, I think inheritance is still better option for this case. So marking this as resolved.

nefeline replied:
> After consideration, I think inheritance is still better option for this case. So marking this as resolved.

Do you mind sharing more info about what made you change your mind here @dinhtungdu ? Even though traits are specifically designed for code-reuse (and that's why I agreed with this approach here in the first place), IMO inheritance and traits are just different mechanisms for achieving the same goals and perhaps more a matter of personal preference as there are no significant improvements in opting for one over another in this case.

dinhtungdu replied:
@nefeline Oh, I'm sorry I missed your [previous comment](https://github.com/woocommerce/woocommerce-blocks/pull/8442#discussion_r1109777286). I changed my mind because:
- for both classes, we need to hook into the WordPress hook system (init method). We might want to add more classes, so duplicating the init method among classes is a code smell to me.
- IMO, inheritance is bad when it becomes a chain. In this case, we only have one level of inheritance, so inheritance becomes more attractive than `trait` to me. We should also mark classes that extend `AbstractTemplateCompatibility` as `final` to ensure that convention.

What do you think?

nefeline replied:
> @nefeline Oh, I'm sorry I missed your https://github.com/woocommerce/woocommerce-blocks/pull/8442#discussion_r1109777286.

No worries & thanks for sharing more details!

> for both classes, we need to hook into the WordPress hook system (init method). We might want to add more classes, so duplicating the init method among classes is a code smell to me.

Agreed. On the other hand, we could also have a trait that doesn't include the init method.

> IMO, inheritance is bad when it becomes a chain. In this case, we only have one level of inheritance, so inheritance becomes more attractive than trait to me.

I'm also not a huge fan of inheritance for this same reason (when we end up with this chain of classes and have to chase down the parent). 

> We should also mark classes that extend AbstractTemplateCompatibility as final to ensure that convention.

My only "but" here would be: making sure every single new PR opened follows this convention and adds the `final` keyword can be easy to miss, whereas with a trait we wouldn't have to worry about that. Either way, I really don't have a strong preference in this case.

dinhtungdu reviewed (commented):
Thank you so much for your work on this PR!

Initially, the `BlockTemplatesCompatibility` is meant for both Archive and Single Product templates. But looking at the complexity of the `SingleProductTemplateCompatibility` class, I can understand why you created an abstraction and a dedicated class for the single product template.

I tested adding another block at the end of the page, after the footer, and the `woocommerce_after_single_product` is injected two times:

<img width="1912" alt="image" src="https://user-images.githubusercontent.com/5423135/219597245-ce5d09a4-3a76-4da2-9c0e-100ef569c3bc.png">

Moreover, looking at the complexity of this PR and #8364:
- In #8364, we added a hidden wrapper to the output of the blocks, which still, is a wrapper. From my test, there is another wrapper also added to the Product Image Gallery at the bottom.
- We're giving merchants more freedom, so it's expected to have blocks added outside the "normal area".
- We create the abstraction for other templates that need the compatibility layer. What are they? If we only need the compatibility layer for archive and single templates, then is an abstraction a bit overkill for the purpose?
- Is it advisable (thank @sunyatasattva for this phrase) to rethink our approach on the wrapper?

### Continue on the wrapper

I have been a supporter of the non-wrapper approach since the meetup. But for a simpler, and more importantly, more reliable backward compatibility layer, a wrapper is super helpful here. Like in #8308, I put all blocks inside the Products block, which enables us to fully replicate the hooks layout of the archive template.

We don't need to choose between wrapper or non-wrapper, we can choose both, which means the block should work inside or outside the wrapper block. But the compatibility layer is only applied to the wrapper and its inner blocks. The wrapper block for the single template in my imagination can be just a variation of the Group block.

dinhtungdu commented on src/Templates/BlockTemplatesCompatibility.php:
```diff
@@ -8,7 +8,7 @@
  * @internal
  */
-class BlockTemplatesCompatibility {
+class BlockTemplatesCompatibility extends AbstractTemplateCompatibility {
```
We should rename this class to remind more about what it does, like `ArchiveProductTemplateCompatibility`. But still, the best case IMO is we can use one class for both template types.

gigitux replied:
I agree with you! I didn't rename it in this PR to avoid confusion. I will create another PR to rename this class when we merge this PR.

gigitux committed:
fix check

gigitux commented:
Thanks for your review. 🙇 

> I tested adding another block at the end of the page, after the footer, and the woocommerce_after_single_product is injected two times:

I fixed it. A check was wrong 6996ca8d54438ac69f60ff7b207b7f2be77b6cc7.

> In https://github.com/woocommerce/woocommerce-blocks/pull/8364, we added a hidden wrapper to the output of the blocks, which still, is a wrapper. From my test, there is another wrapper also added to the Product Image Gallery at the bottom.

Yep, it works as expected. There are the details of this decision in the code, but you can find more context in this discussion: https://github.com/woocommerce/woocommerce-blocks/pull/8364#pullrequestreview-1282916377

We are adding this wrapper only on the front end. For the merchant, on the editor side, this wrapper doesn't exist. Furthermore, this solution is to help with the transition. I can imagine that in the next months, we can offer the possibility to disable the "compatibility layer".

> We're giving merchants more freedom, so it's expected to have blocks added outside the "normal area".

What do you mean?

> We create the abstraction for other templates that need the compatibility layer. What are they? If we only need the compatibility layer for archive and single templates, then is an abstraction a bit overkill for the purpose?

Putting everything in one class would create too much confusion. Furthermore, the Rubik team is exploring the templates (https://github.com/woocommerce/woocommerce-blocks/pull/8350), so in the future, we could have a compatibility layer for multiple templates. Having a common interface is the best way to have a standard solution to resolve this issue, IMHO.

> We don't need to choose between wrapper or non-wrapper, we can choose both, which means the block should work inside or outside the wrapper block. But the compatibility layer is only applied to the wrapper and its inner blocks. The wrapper block for the single template in my imagination can be just a variation of the Group block.

If we want to follow Gutenberg's philosophy, we should not use the wrapper (per0F9-b5-p2). Also, as I said above, ideally, the "wrapper" should be just a temporary solution to help the transition without breaking the old extension.

nefeline commented on src/Templates/SingleProductTemplateCompatibility.php:
```diff
@@ -0,0 +1,413 @@
+	 * @return bool
+	 */
+	private function is_single_product_template() {
+		return is_product();
```
I'd recommend ditching this `is_single_product_template()` method and relying directly on `is_product()`. Additionally, this function is unaware of any templates; it will return true whenever you visualize a single post with the type = product.

gigitux replied:
> I'd recommend ditching this is_single_product_template() method and relying directly on is_product().

👍 

> Additionally, this function is unaware of any templates; it will return true whenever you visualize a single post with the type = product.

Right, but according to WP hierarchy, if you are visualizing a single post with type product, you are visualizing the Single Product Template, no?

nefeline replied:
> Right, but according to WP hierarchy, if you are visualizing a single post with type product, you are visualizing the Single Product Template, no?

Correct, but just as a coincidence, not because `is_product()` is verifying the template's name. A single post on WP can have multiple templates (as we are already in the process of implementing for each one of the product types, for example).

gigitux replied:
WC core uses this function to check if it is loaded a Single Product page ([source code](https://github.com/woocommerce/woocommerce/blob/3681b84a4c524691ad0748d29992a2d75df87fde/plugins/woocommerce/includes/class-wc-frontend-scripts.php/#L402-L416)).

> As we are already in the process of implementing for each one of the product types, for example.

Agree and this was the reason that I created a dedicated function. I'm pretty sure that we will have to revisit this logic when the product types template is implemented.

nefeline replied:
> WC core uses this function to check if it is loaded a Single Product page ([source code](https://github.com/woocommerce/woocommerce/blob/3681b84a4c524691ad0748d29992a2d75df87fde/plugins/woocommerce/includes/class-wc-frontend-scripts.php/#L402-L416)).

Yep! My only observation here is regarding the name of the method in particular: as `is_single_product_template()` was "wrapping/renaming" `is_product()`,  and that's not a valid assumption because `is_product` is only checking for the existence of a single post of a type = 'product'. Thanks for addressing the feedback!

nefeline commented on src/Templates/SingleProductTemplateCompatibility.php:
```diff
@@ -0,0 +1,413 @@
+	 * For more details, see https://github.com/woocommerce/woocommerce-blocks/issues/8314.
+	 *
+	 * @param string $template_content Template Content.
+	 * @return string Wrapped template content inside a div.
```
the returned value by `array_map` is an array, not a string.

gigitux replied:
Nice catch! Thanks!

nefeline commented on src/Templates/SingleProductTemplateCompatibility.php:
```diff
@@ -0,0 +1,413 @@
+				$block          = $item[0];
+
+				if ( 'core/template-part' === $block['blockName'] ) {
+					array_push( $carry['template'], $block );
```
nit: since just one element is being pushed to the array, could use `$carry[] = array( $block );` instead (same for all usages of array_push in this method).

gigitux replied:
Thanks for raising this! I didn't know about this. Also, I'm not a huge fan of this syntax, but it is used in our codebase. I will update the code 👍

nefeline commented on src/Templates/AbstractTemplateCompatibility.php:
```diff
@@ -0,0 +1,157 @@
+				continue;
+			}
+			foreach ( $data['hooked'] as $callback => $priority ) {
+				remove_action( $hook, $callback, $priority );
```
Is it correct to assume this is a temporary measure until we stop relying on the backend for rendering these blocks? I can imagine performance won't be optimal until we reach this point, considering we are removing the functions attached to these hooks to re-add them next.

gigitux replied:
I'm tagging @dinhtungdu that he writes this part of the code.

> Is it correct to assume this is a temporary measure until we stop relying on the backend for rendering these blocks?

We will keep relying on the backend to render these blocks. It would be interesting to see how much overhead we are adding with this compatibility layer, but the real long-term is to remove this latter 🚀

nefeline replied:
> We will keep relying on the backend to render these blocks. It would be interesting to see how much overhead we are adding with this compatibility layer, but the real long-term is to remove this latter 🚀

Cool! I asked because if we are planning to rely on this in the long term, we could consider making some changes to the core of Woo, so these functions are not hooked in the first place if the template in use is one of the new options we are introducing here.

dinhtungdu replied:
> Is it correct to assume this is a temporary measure until we stop relying on the backend for rendering these blocks?

Not really. It will be there even after we stop relying on the backend rendering for those blocks. We will remove the whole layer when the extensions ecosystem is ready for block themes.

> I can imagine performance won't be optimal until we reach this point, considering we are removing the functions attached to these hooks to re-add them next

@nefeline we won't re-add them later. We only remove default hooks added by WC Core because we have equivalent blocks in the templates, so there is no need to add them back. Btw, in terms of performance, I think removing hooks is pretty cheap.

> I asked because if we are planning to rely on this in the long term, we could consider making some changes to the core of Woo, so these functions are not hooked in the first place if the template in use is one of the new options we are introducing here.

We can always update WC Blocks when WC Core changes its default hooks handling.

nefeline replied:
> Not really. It will be there even after we stop relying on the backend rendering for those blocks. We will remove the whole layer when the extensions ecosystem is ready for block themes.

Gotcha! Thanks for clarifying.

> we won't re-add them later. We only remove default hooks added by WC Core because we have equivalent blocks in the templates, so there is no need to add them back. Btw, in terms of performance, I think removing hooks is pretty cheap.

Yep, that's what I meant with re-adding them later: because in most cases (if not all of them), we are relying on the same template functions as the core templates are on Woo, so when a block is inserted, we are hooking the same methods/functions back.

dinhtungdu replied:
> so when a block is inserted, we are hooking the same methods/functions back.

Can you clarify your point more? I'm not sure I get this right.

We only prepend/append the hook (`do_action` calls) to the output of some blocks, we don't 'hook' any default methods added by WC Core (that are removed) back. Most of our blocks for archive/single product templates are now powered by WC template functions, but we use those functions directly in the render method of blocks, not hooking them back.

nefeline replied:
> Can you clarify your point more? I'm not sure I get this right.

Sure! Taking the breadcrumbs block as an example. If we follow the order of execution, we have:

1. First, the core of Woo hooks `woocommerce_breadcrumb` to `'woocommerce_before_main_content'`:

[`add_action( 'woocommerce_before_main_content', 'woocommerce_breadcrumb', 20, 0 );`](https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce/includes/wc-template-hooks.php#L47)

2. Next, if the breadcrumb block is used, we unhook this function via [`remove_default_hooks`](https://github.com/woocommerce/woocommerce-blocks/blob/ce9bd9cc9c676b8e5d191e7975133b530f322e4f/src/Templates/AbstractTemplateCompatibility.php#L92).

3. Last but not least, the [`woocommerce_breadcrumb`](https://github.com/woocommerce/woocommerce-blocks/blob/49ce2401047d87efdc7b6a3cbc516e3de70386f8/src/BlockTypes/Breadcrumbs.php#L30) function is then invoked by the block.

What I was wondering is if we can add a conditional on step 1 (something as `is_block_in_use( 'breadcrumb' )`, in the core of Woo) and remove the need for step 2 altogether. What do you think?

nefeline replied:
Actually, as a second thought: I believe this approach makes more sense only when Woo blocks start being the norm, not the exception (in other words, when the majority of the online stores are using blocks instead of classic templates): adding an `if` condition on core at this point might be a bit of an overhead. Marking as resolved: thanks for the discussion!

nefeline reviewed (commented):
Nice work 👏 ! It's interesting to see how we handle hooks for the templates as we slowly transition to a "fully blockified" experience! I've left a few comments/suggestions.

gigitux committed:
address feedback

gigitux commented:
@nefeline, thanks for the review 🙇 
I addressed your feedback with the commit 7177481b6e2d0e29f2e6c4eb70ccae19a35f9bd3 (except for using the trait, as I said above, I want to do it in a dedicated PR)

nefeline commented:
> @nefeline, thanks for the review 🙇
I addressed your feedback with the commit https://github.com/woocommerce/woocommerce-blocks/commit/7177481b6e2d0e29f2e6c4eb70ccae19a35f9bd3 (except for using the trait, as I said above, I want to do it in a dedicated PR)

Thank you so much; this is looking sharp! Regarding the inheritance x trait approach: at the end of the day, they are different mechanisms for achieving the same goals. I've left a [FU note](https://github.com/woocommerce/woocommerce-blocks/pull/8442#discussion_r1111212805) to our discussion, but not a blocker as I believe these are interchangeable solutions.

dinhtungdu commented on src/Templates/AbstractTemplateCompatibility.php:
```diff
@@ -0,0 +1,157 @@
+	 */
+	protected $hook_data;
+
+	/**
+	 * Constructor.
+	 */
+	public function __construct() {
+		$this->set_hook_data();
+		$this->init();
+	}
+
+	/**
+	 * Initialization method.
+	 */
+	protected function init() {
+		if ( ! wc_current_theme_is_fse_theme() ) {
+			return;
+		}
+
+		add_filter( 'render_block_data', array( $this, 'update_render_block_data' ), 10, 3 );
+		add_filter( 'render_block', array( $this, 'inject_hooks' ), 10, 2 );
+	}
```
This is my mistake from the initial implementation. We shouldn't add hooks in the constructor of the class. Instead, we should fire the `init` method in the `Bootstrap` class instead.

```suggestion
	/**
	 * Initialization method.
	 */
	protected function init() {
		if ( ! wc_current_theme_is_fse_theme() ) {
			return;
		}

		$this->set_hook_data();

		add_filter( 'render_block_data', array( $this, 'update_render_block_data' ), 10, 3 );
		add_filter( 'render_block', array( $this, 'inject_hooks' ), 10, 2 );
	}
```

dinhtungdu commented on src/Templates/BlockTemplatesCompatibility.php:
```diff
@@ -31,18 +31,6 @@ public function __construct() {
 		$this->init();
 	}
```
The `$hook_data` property declaration and `__construct` method of this class should be removed.

gigitux replied:
Fixed via 4c888cf7395c090da1052c8bc2c966a2f6e59c49

dinhtungdu commented on src/Domain/Bootstrap.php:
```diff
@@ -129,6 +130,7 @@ function() {
 		$this->container->get( ProductAttributeTemplate::class );
 		$this->container->get( ClassicTemplatesCompatibility::class );
 		$this->container->get( BlockTemplatesCompatibility::class );
+		$this->container->get( SingleProductTemplateCompatibility::class );
```
```suggestion
		$this->container->get( BlockTemplatesCompatibility::class )->init();
		$this->container->get( SingleProductTemplateCompatibility::class )->init();
```

gigitux replied:
Fixed via 4c888cf7395c090da1052c8bc2c966a2f6e59c49

dinhtungdu commented on src/BlockTemplatesController.php:
```diff
@@ -4,6 +4,7 @@
 use Automattic\WooCommerce\Blocks\Domain\Package;
 use Automattic\WooCommerce\Blocks\Templates\BlockTemplatesCompatibility;
```
This is unused in favor of `SingleProductTemplateCompatibility`
```suggestion
```

gigitux replied:
Fixed via 10bc1b15c09a7d82e067c1c68befe0ecd57981fa

dinhtungdu commented on src/Templates/SingleProductTemplateCompatibility.php:
```diff
@@ -0,0 +1,402 @@
+		return $wrapped_blocks;
+	}
+
+
```
Double empty line here.
```suggestion
```

gigitux replied:
Fixed via 512bdab53ba75a82f60abb6cc81984bba2fa9b3e

dinhtungdu reviewed (commented):
I added some more comments to fix my mistake from the initial implementation.

dinhtungdu reviewed (commented):
@gigitux Can you also help me change the visibility of these methods in `BlockTemplatesCompatibility` from `protected` to `private`? 🙇🏽‍♂️ 

- is_archive_template
- inner_blocks_walker
- inject_attribute

gigitux committed:
remove unused import

gigitux committed:
double empty line

gigitux committed:
remove logic in the constructor

gigitux commented:
> @gigitux Can you also help me change the visibility of these methods in `BlockTemplatesCompatibility` from `protected` to `private`? 🙇🏽‍♂️
> 
> * is_archive_template
> * inner_blocks_walker
> * inject_attribute

Thanks for the review! I addressed your feedback!

As I said above, I will rename the class to `ArchiveProductTemplateCompatibility ` in a dedicated PR.

gigitux committed:
Merge branch 'trunk' into add/compatibility-single-product-block

gigitux committed:
Merge branch 'trunk' into add/compatibility-single-product-block

nefeline commented on src/Templates/SingleProductTemplateCompatibility.php:
```diff
@@ -0,0 +1,401 @@
+				'position'   => 'before',
+				'hooked'     => array(
+					'woocommerce_output_all_notices' => 10,
+					'woocommerce_breadcrumb'         => 20,
```
The breadcrumb template function is hooked to `'woocommerce_before_main_content'` rather than `'woocommerce_before_single_product'`

https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce/includes/wc-template-hooks.php#L47

gigitux replied:
Nice catch! Fixed via [ebe5b5e](https://github.com/woocommerce/woocommerce-blocks/pull/8442/commits/ebe5b5e825798eeb8b004fcb693f29f6e54eb71b)

nefeline replied:
Thanks for the update! Do you think it's worth adding a compatibility layer for the breadcrumb elsewhere (or maybe just add the correct hook `'woocommerce_before_main_content'` to this array, even though it is not strict to the single product template)?

gigitux replied:
Mmm, why? The breadcrumb is a wrapper of the PHP code, so it already support all the hooks, no?

nefeline replied:
> Mmm, why? The breadcrumb is a wrapper of the PHP code, so it already support all the hooks, no?

Yep! TBH I'm not sure if this is necessary or not, as I'm still not super familiar with the background decisions made for this compatibility layer. What I was wondering is if, the same way we had the `'woocommerce_breadcrumb'` hooked here under `'woocommerce_before_single_product'`, we wouldn't want to keep the same pattern and just move it to `'woocommerce_before_main_content'` instead (which is the correct one)? Maybe it's not a requirement now that it is not under `'woocommerce_before_single_product'` anymore, just wanted to double-check.

gigitux replied:
All **the callbacks** that we put inside `hooked` are removed because we don't want that WooCommerce injects element in the DOM because now we have the blocks.

For this reason, we have this bug https://github.com/woocommerce/woocommerce-blocks/issues/8528 because we are removing the callback `woocommerce_show_product_sale_flash`. In this case, we should keep this callback (I will create a dedicated PR for this).

Does it make sense? Feel free to reach me if you have other questions!

nefeline replied:
> All the callbacks that we put inside hooked are removed because we don't want that WooCommerce injects element in the DOM because now we have the blocks.

Thanks, but that wasn't really my question. My question is regarding the breadcrumbs: because [we were previously removing its callback](https://github.com/woocommerce/woocommerce-blocks/pull/8442#discussion_r1113458023) from `'woocommerce_before_single_product'` (which was incorrect ), and now it was [simply removed with no replacement](https://github.com/woocommerce/woocommerce-blocks/pull/8442/commits/ebe5b5e825798eeb8b004fcb693f29f6e54eb71b) (which I believe there should be).

gigitux replied:
Ah, gotcha! In 84d03e0f8936d789f6a26b4c00e79550e3e5d720, I added the missing hooks 💪

nefeline replied:
Just to better illustrate, I'm talking about something along these lines:
```
			'woocommerce_before_main_content'         => array(
				'block_name' => '',
				'position'   => 'before',
				'hooked'     => array(
					'woocommerce_breadcrumb' => 20,
				),
			),
```

nefeline replied:
> Ah, gotcha! In https://github.com/woocommerce/woocommerce-blocks/commit/84d03e0f8936d789f6a26b4c00e79550e3e5d720, I added the missing hooks 💪

Cool, thanks! I'm gonna check & re-run the tests soon.

gigitux committed:
remove hook

gigitux committed:
Merge branch 'add/compatibility-single-product-block' of https://github.com/woocommerce/woocommerce-blocks into add/compatibility-single-product-block

gigitux committed:
Merge branch 'trunk' into add/compatibility-single-product-block

gigitux committed:
Merge branch 'trunk' into add/compatibility-single-product-block

gigitux committed:
Merge branch 'trunk' into add/compatibility-single-product-block

gigitux committed:
Merge branch 'trunk' of https://github.com/woocommerce/woocommerce-blocks into add/compatibility-single-product-block

gigitux committed:
generate the docs

gigitux committed:
add missing hooks

gigitux committed:
fix docs

Aljullu commented on src/Templates/AbstractTemplateCompatibility.php:
```diff
@@ -0,0 +1,157 @@
+		 * @since 9.5.0
+		 * @param array $data Additional hooked data. Default to empty
+		 */
+		$additional_hook_data = apply_filters( 'woocommerce_blocks_hook_compatibility_additional_data', array() );
```
I'm not sure if I'm understanding correctly what this filter does, but I find the name a bit confusing. Would something like `woocommerce_blocks_hooks_to_skip_in_blocks` better describe what it does? (Probably there are better names, but just trying to see if I understand what it does)

In case we aren't sure about the name, I would mark it as `__experimental` so we can iterate on it in the future.

gigitux replied:
This filter has been introduced by [this PR](https://github.com/woocommerce/woocommerce-blocks/pull/8172/files#diff-02e3c036540a6c7806ce8f3ca2cb1bbcd3d1d1ce67400012c23dd066f87786a9R296).

I'm okay with updating the name, but we already released a version with this hook.

Aljullu replied:
Oh, my bad.  :sweat_smile:  Let's move ahead with this name, then.

Aljullu commented on src/Templates/SingleProductTemplateCompatibility.php:
```diff
@@ -0,0 +1,425 @@
+			return sprintf(
+				'%1$s%2$s%3$s',
+				$this->get_hooks_buffer(
+					$first_block_hook['before'],
```
Should we merge the first and last block's `before` hooks? You are doing so with the [`after` ones](https://github.com/woocommerce/woocommerce-blocks/pull/8442/files#diff-bbdd2400423027ed235c89d88361c9e5e2a4a6e4d28e6282b97a67d406dcdbd8R96-R102).
```suggestion
					array_merge(
						$first_block_hook['before'],
						$last_block_hook['before']
					),
```

gigitux replied:
I will update it, but `$last_block_hook['before']` is empty.

Edit: Oh ok, you are aware of this :D

Aljullu commented on src/Templates/SingleProductTemplateCompatibility.php:
```diff
@@ -0,0 +1,425 @@
+		if ( isset( $block['attrs'][ self::IS_LAST_BLOCK ] ) ) {
+			return sprintf(
+				'%1$s%2$s%3$s',
+				$this->get_hooks_buffer( array(), 'before' ),
```
Similar comment, shouldn't this be:
```suggestion
				$this->get_hooks_buffer(
					$last_block_hook['before'],
					'before'
				),
```

(I know `$last_block_hook['before']` is empty, but I find it confusing that we are declaring it but not using it anywhere :confused: )

Aljullu commented on src/Templates/SingleProductTemplateCompatibility.php:
```diff
@@ -0,0 +1,425 @@
+		$parsed_blocks  = parse_blocks( $template_content );
+		$grouped_blocks = self::group_blocks( $parsed_blocks );
+
+		// @TODO: Check this list before terminating the Blockfied Single Product Template project.
```
We usually use `@todo` (lowercase and no `:`) in todo comments. Example: https://github.com/woocommerce/woocommerce-blocks/blob/1e5f5a03ce90410998023b7ca00dfa367d638d50/assets/js/types/type-defs/currency.ts#L11

Aljullu commented on src/Templates/SingleProductTemplateCompatibility.php:
```diff
@@ -0,0 +1,425 @@
+					return $carry;
+				}
+
+				if ( false === $carry['first_block']['found'] ) {
```
Not a blocker at all, but I think the code could be simplified if we removed the `found` attribute and instead checked whether `index` is not empty.

Aljullu reviewed (approved):
Amazing job with this PR! :clap: Really nice to see how this is shaping up.

I'm pre-approving since it looks good to me, but I left some comments below (at least the filter name one, I think it's a blocker).

gigitux committed:
Merge branch 'trunk' of https://github.com/woocommerce/woocommerce-blocks into add/compatibility-single-product-block

gigitux committed:
address feedback

gigitux committed:
Merge branch 'trunk' of https://github.com/woocommerce/woocommerce-blocks into add/compatibility-single-product-block

gigitux committed:
fix linter

gigitux committed:
Merge branch 'trunk' of https://github.com/woocommerce/woocommerce-blocks into add/compatibility-single-product-block

gigitux committed:
fix import

gigitux committed:
Disable compatibility layer when the WooCommerce Product Grid Block block and WooCommerce Single Product Block are used (#8538)

* disable compatibility layer via hook

* update docs

* generate the docs

* fix version

* fix import

* fix code after merge

gigitux committed:
Merge branch 'trunk' of https://github.com/woocommerce/woocommerce-blocks into add/compatibility-single-product-block

gigitux commented:
> Since this PR is already big, I plan to merge this when you approve and, after that, in another PR, rename BlockTemplatesCompatibility to ArchiveProductTemplateCompatibility and convert this class to a trait.

In the end, we decided to continue to use a class. I renamed the class with #8579.
